### PR TITLE
doc: add ncs-include sphinx extension

### PR DESCRIPTION
**Pull request by utzig**
This extension behaves similarly to ".. include", but with the following added options:

* docset (str): This is a documentation set directory, which must available in `ncs_include_mapping` on conf.py. `ncs_include_mapping` is a dict with a docset key and the documentation set (or any directory under it) path.
* indent (int): add the given amount of spaces to the beginning of every imported line.
* dedent (int): remove the given amount of spaces to the beginning of every line.
* auto-dedent (flag): when set tries to remove enough spaces from the beginning of every line so they are "stripped" of left spaces, and should match the formatting of the scope around.

An example was added importing Window required tooling from Zephyr.

### CODE REVIEW

File: doc/nrf/conf.py
Comment (tejlmand) [inline doc/nrf/conf.py:88]: could `ncs_include` be a too generic name, that we could want to use for other python packages in future ?

Comment (carlescufi) [reply doc/nrf/conf.py:88]: but this doesn't collide with python packages, because the python package, if uploaded to pip is always prefixed with`sphinx-` or `sphinxcontrib-`

Comment (carlescufi) [reply doc/nrf/conf.py:88]: What you see here is the name of the extension, not the package. @utzig can confirm of course.

Comment (utzig) [reply doc/nrf/conf.py:88]: Correct, it's the extension's name, I suppose it could be generalized and useful for other projects, but I had no intention of publishing it as a package since it was tailored for NCS, and I am not sure if anyone else uses a similar setup.

Comment (tejlmand) [reply doc/nrf/conf.py:88]: I know this is not a package.
I'm not thinking on publishing this one, more in case there is a second `ncs_include` package in future.

I don't know exactly the code for sphinx, but I assume it imports all extensions from this `extensions` variable.
And I do see that we have:
```
sys.path.append(os.path.join(NRF_BASE, 'scripts', 'sphinx_extensions'))
```
so in case there was ever to be second `ncs_include` package installed in the system, I believe there could be a risk of collision.

So I just like to play it safe :wink:

**Review by tejlmand — APPROVED** [review]
code looks fine

**Comment by utzig** [comment]
Before I forget it, commit cc9db74f5d1b13b3aa0c4a4795ea395d2e8637e9 was added only as a demonstration based on a previous example, I am not sure it should be merged as part of this PR once it is approved; @ru-fu let me know if I should drop it.

File: scripts/sphinx_extensions/ncs_include/ncs_include.py
Comment (ru-fu) [inline scripts/sphinx_extensions/ncs_include/ncs_include.py:41]: Since docset is a required option - can we define "nrf" as a docset as well, so we can take advantage of the indent/dedent options inside our main doc set as well?
Or even better, could the "local" doc set be the default?

**Review by ru-fu — COMMENTED** [review]
Thank you, this looks great!
Running some tests now. :)

**Comment by ru-fu** [comment]
> 
> 
> Before I forget it, commit [cc9db74](https://github.com/nrfconnect/sdk-nrf/commit/cc9db74f5d1b13b3aa0c4a4795ea395d2e8637e9) was added only as a demonstration based on a previous example, I am not sure it should be merged as part of this PR once it is approved; @ru-fu let me know if I should drop it.

Yes, the commit should be dropped, since we need to clean up this topic before we can include the content. I copied the info on how to do it into our Jira task for cleaning up the topic. :)

**Comment by ru-fu** [comment]
In this case, auto-dedent messes up the output and causes a warning:
```
#. Follow this guide to:

   .. ncs-include:: getting_started/index.rst
         :docset: zephyr
         :auto-dedent:
         :start-after: Follow this guide to:
         :end-before: _host_setup
```
It works without the :auto-dedent option though.

**Comment by utzig** [comment]
> In this case, auto-dedent messes up the output and causes a warning:
> 
> ```
> #. Follow this guide to:
> 
>    .. ncs-include:: getting_started/index.rst
>          :docset: zephyr
>          :auto-dedent:
>          :start-after: Follow this guide to:
>          :end-before: _host_setup
> ```
> 
> It works without the :auto-dedent option though.

Please push your test to a branch so I can take a look at what's happening.

**Comment by ru-fu** [comment]
> Please push your test to a branch so I can take a look at what's happening.

https://github.com/ru-fu/sdk-nrf/commit/c56192580dca6011ab2544b3a80e30e85fef8c74

File: scripts/sphinx_extensions/ncs_include/ncs_include.py
Comment (utzig) [reply scripts/sphinx_extensions/ncs_include/ncs_include.py]: I both defined "nrf" and added code for using the local doc set when no "docset" was given.

**Comment by utzig** [comment]
> > Please push your test to a branch so I can take a look at what's happening.
> 
> [ru-fu@c561925](https://github.com/ru-fu/sdk-nrf/commit/c56192580dca6011ab2544b3a80e30e85fef8c74)

The bug should be fixed. Unfortunately this masked an issue with importing the "chocolatey" tips from Zephyr, because the link specifications are not indent (0 indentation) which breaks the heuristic. In my commit here, I broke the import into two "ncs-include" steps, which works, but I am not sure on how it can be fixed to run in a single step (if that is deemed required!).

File: doc/nrf/conf.py
Comment (utzig) [reply doc/nrf/conf.py:56]: I think it would be better to create a "nrf" (or "ncs") namespace, like Zephyr also does for its extensions, and include this one, "boards_from_rows", "kconfig_from_options", "interbreathe" (which I am not sure is still required with latest Sphinx), and possibly others. But I sure don't want to move things around here! :-)

**Comment by ru-fu** [comment]
> The bug should be fixed. Unfortunately this masked an issue with importing the "chocolatey" tips from Zephyr, because the link specifications are not indent (0 indentation) which breaks the heuristic. In my commit here, I broke the import into two "ncs-include" steps, which works, but I am not sure on how it can be fixed to run in a single step (if that is deemed required!).

Would it be possible to update the :dedent: option to dedent **max** the given amount of spaces? Then we could use it in cases like these and say that we want to dedent 6 spaces, but for those lines where there are less leading spaces, it just removes all of them.

But this is probably not the most common problem, so if this is a lot of work, let's leave it.

**Review by ru-fu — COMMENTED** [review]
Functionality looks good, but one more suggestion (if it is feasible without a huge effort).

Please remember to drop the test commit before this is merged.

**Comment by utzig** [comment]
@ru-fu I updated `:dedent:` to remove up to the given number of spaces, so in the Zephyr's "chocolatey" example, it's possible to use `:dedent: 6` and it will remove 6 spaces or less (if the line has less) from the beginning, so it will properly skip lines that are non-indented like the replace directive lines.

**Comment by utzig** [comment]
> Please remember to drop the test commit before this is merged.

Done.

**Review by ru-fu — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Pull request merged**
